### PR TITLE
Use request context from Chime Atlas

### PR DESCRIPTION
**Pull request by ziad-abdo**
For capturing and surfacing request_id and other metadata.

### CODE REVIEW

File: lib/simplekiq/metadata_client.rb [hunk 1]
```diff
@@ -16,15 +17,10 @@ def call(_worker, job, _queue, *)
     end
 
     def client_metadata
-      {
+      Chime::Atlas::RequestContext.current.to_h.stringify_keys.merge(
```
Comment (abernard) [inline lib/simplekiq/metadata_client.rb]: I think we should have a default behavior here that is safe if `Chime::Atlas::RequestContext` does not exist.

File: lib/simplekiq/metadata_server.rb [hunk 2]
```diff
@@ -58,11 +59,11 @@ def processed_by_host
...
     end
 
-    def add_request_id_to_thread(job)
-      return if job['request_id'].nil?
+    def apply_metadata_to_thread(job)
+      metadata = job.symbolize_keys.slice(:request_id, :trace_id, :sampling_priority)
 
       # In the event the current job enqueues other jobs it will use the same request_id
-      Thread.current['atlas.request_id'] = job['request_id']
+      Chime::Atlas::RequestContext.set(metadata)
```
Comment (abernard) [inline lib/simplekiq/metadata_server.rb]: Same safe behavior here.

File: spec/lib/simplekiq/metadata_server_spec.rb [hunk 3]
```diff
@@ -8,6 +8,13 @@ def perform(_)
   end
 end
 
+class ThreadCleaner
+  def call(_worker, _job, _queue, *)
+    Chime::Atlas::RequestContext.clear
```
Comment (abernard) [inline spec/lib/simplekiq/metadata_server_spec.rb]: and here.

**Comment by abernard** [comment]
I agree with @fzf .  I would propose we create a dummy method that can provide a payload, but just returns an empty hash.  We can then use a DI style monkey patch in all services, performed as a Rails initializer.  It could use Atlas.

File: lib/simplekiq/metadata_client.rb (hunk 1 above)
Comment (dalyons) [reply lib/simplekiq/metadata_client.rb]: how? its not going to not exist. atlas is a dependency. 
simplekiq is not going to be used outside of chime ruby projects, which will have atlas.

Comment (abernard) [reply lib/simplekiq/metadata_client.rb]: Very good point.  My mental model went directly to @fzf's comment before about not including Atlas.  I had attempted this once before https://github.com/1debit/simplekiq/pull/23#discussion_r279113268 and I think his comment is still relevant.

There were various behaviors that we wanted consistency for in simplekiq, but I don't think pulling in the kitchen sink for one field here is necessary.

**Comment by dalyons** [comment]
Simplekiq is only going to be used by chime ruby projects, as its effectively "chime sidekiq".  Said projects are always going to have atlas. There's no real downside to including atlas as a dependency. 

Furthermore, the entire point of centralizing all request context in atlas was to put it behind a common interface, and remove "magic at a distance" (the old way of reading a string-key'd thread local hoping it had been set somewhere upstream)  . Doing a "DI style monkey patch in all services, performed as a Rails initializer" sounds like it would be super indirect and very hard to follow how/where request context gets passed, undoing that clarity. But maybe i dont understand what you have in mind there.

In my ideal world these middlewares would be moved entirely to `atlas/sidekiq` - just like we have an `atlas/rack` & `atlas/faraday` for their respective chime-flavoured middlewares. And then the app could instantiate/configure  a "standard chime sidekiq" from atlas, in the same way we currently use it to configure a "standard chime rails/rack" and a "standard chime client connection (faraday)".

File: lib/simplekiq/metadata_client.rb (hunk 1 above)
Comment (dalyons) [reply lib/simplekiq/metadata_client.rb]: It’s three fields already, not one, and I plan on adding at least two more (“user Id and is_office”)
The whole point of doing it this way is nothing will have to change in simplekiq when I add those, because it’s behind the interface

**Comment by abernard** [comment]
I think because said projects will always have Atlas is as good of a reason as any to use DI and not pull Atlas in.

Pulling in Atlas risks irritating transitive dependency issues.  Every service will also always run Atlas' initializer.  Putting the request context initializer in there guarantees that any metadata type changes are independent of sidekiq and don't need 2 gem updates x N services.  If one day we wanted to have `user_id` in there for some reason, that would also be independent of this system.

I would suggest that Atlas is more fundamental to the metadata story than this async processing library.  Queues, RPC service meshes, logging etc. could take advantage of that information, and I see no reason to replicate that in here if we don't have to.

**Comment by dalyons** [comment]
> Pulling in Atlas risks irritating transitive dependency issues. Every service will also always run Atlas' initializer. 

they already do.

> If one day we wanted to have `user_id` in there for some reason, that would also be independent of this system.

yes, thats what im saying, and why it uses an explicit interface for dumping/loading the context; so it can change arbitarily. 

> I would suggest that Atlas is more fundamental to the metadata story than this async processing library. Queues, RPC service meshes, logging etc. could take advantage of that information, and I see no reason to replicate that in here if we don't have to.

i dont know what you mean by "replicate it in here"; we're calling into the library that does it.  


standing back, I guess I dont see the reason for these "chime specific configurations" of libraries (rack/faraday/sidekiq/etc) to be separate gems. They're never going to be used independently, and they're never going to be used outside of chime, so why suffer the management/maintenance overhead of making them gems instead of just putting them in the "standard chime rails toolkit" (atlas)? 
Then if the consuming app wants "chime sidekiq" they can instantiate one from the atlas sdk.  
But thats a much bigger change i guess.

**Comment by dalyons** [comment]
(i think maybe we both need to demonstrate what we mean with some code-ish tomo, sorry too tired to be interpreting you properly i think)

File: lib/simplekiq/metadata_client.rb (hunk 1 above)
Comment (tilo) [reply lib/simplekiq/metadata_client.rb]: Simplekiq is not published in RubyGems
